crates/dbsp/src/circuit/metrics.rs

```diff
@@ -57,9 +66,290 @@ pub const COMPACTION_DURATION: &str = "file.compaction_duration";
 /// Time a worker was stalled waiting for more merges to complete.
 pub const COMPACTION_STALL_TIME: &str = "file.compaction_stall_time";
 
-/// Number of records dropped due to LATENES annotations
+/// Number of records dropped due to LATENESS annotations
 pub const TOTAL_LATE_RECORDS: &str = "records.late";
 
+/// Runtime in microseconds of an Operator evaluation
+pub const OPERATOR_EVAL_DURATION: &str = "operator.runtime_micros";
+
+/// Holds a map of the [`DBSPMetric`] and their equivalent [`metrics`] type.
+///
+/// [`RwLock`]s here should be very efficient, as reading is the most common operation.
+/// Writing is only necessary while adding a new metric.
+///
+/// Updating the inner type in the map doesn't require mutable access due to the interior
+/// mutability of the [`Counter`], [`Gauge`] and [`Histogram`] types.
+#[allow(unused)]
+#[derive(Default)]
+struct MetricsRecorder {
+    counters: RwLock<HashMap<DBSPMetric, Counter>>,
+    gauge: RwLock<HashMap<DBSPMetric, Gauge>>,
+    histogram: RwLock<HashMap<DBSPMetric, Histogram>>,
+}
+
+lazy_static! {
+    /// Global recorder for DBSP related metrics.
+    static ref METRICS_RECORDER: Arc<MetricsRecorder> = Arc::new(MetricsRecorder::new());
+}
+
+impl MetricsRecorder {
+    fn new() -> Self {
+        Self::default()
+    }
+
+    /// Registers a new [`Counter`] and saves it to the map.
+    #[allow(unused)]
+    fn register_counter(
+        &self,
+        metric: DBSPMetric,
+        unit: Option<MetricUnit>,
+        description: Option<String>,
+    ) {
+        {
+            let c = self.counters.read().unwrap();
+            if c.contains_key(&metric) {
+                return;
+            }
+        }
+
+        metric.describe(MetricType::Counter, unit, description);
+
+        {
+            let mut c = self.counters.write().unwrap();
+            let counter = metrics::counter!(metric.name(), metric.labels.into_labels());
+            c.insert(metric, counter);
+        }
+    }
+
+    /// Increments the [`Counter`] for the given `metric` by `value`.
+    #[allow(unused)]
+    fn increment_counter(&self, metric: DBSPMetric, value: u64) {
```

Comment:
`DBSPMetric` is a very expensive data structure to instantiate on each counter update, since it includes a bunch of dynamically allocated fields. There should be 0 dynamic allocation in the common path of updating a metric.